src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -3104,13 +3106,97 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            }
 
             return objectCreation;
         }
 
+        public override IOperation VisitMemberInitializer(IMemberInitializerOperation operation, int? captureIdForResult)
+        {
+            Debug.Assert(_currentImplicitReceiver != null);
+
+            // We explicitly do not push the initialized member onto the stack here. We visit the initialized member to get the implicit receiver that will be substitued in when an
+            // IInstanceReferenceOperation with InstanceReferenceKind.ImplicitReceiver is encounted. If that receiver needs to be pushed onto the stack, its parent will handle it.
+            // In member initializers, the code generated will evalute InitializedMember multiple times. For example, if you have the following:
+            //
+            // class C1
+            // {
+            //   public C2 C2 { get; set; } = new C2();
+            //   public void M()
+            //   {
+            //     var x = new C1 { C2 = { P1 = 1, P2 = 2 } };
+            //   }
+            // }
+            // class C2
+            // {
+            //   public int P1 { get; set; }
+            //   public int P2 { get; set; }
+            // }
+            //
+            // We generate the following code for C1.M(). Note the multiple calls to C1::get_C2().
+            //   IL_0000: nop
+            //   IL_0001: newobj instance void C1::.ctor()
+            //   IL_0006: dup
+            //   IL_0007: callvirt instance class C2 C1::get_C2()
+            //   IL_000c: ldc.i4.1
+            //   IL_000d: callvirt instance void C2::set_P1(int32)
+            //   IL_0012: nop
+            //   IL_0013: dup
+            //   IL_0014: callvirt instance class C2 C1::get_C2()
+            //   IL_0019: ldc.i4.2
+            //   IL_001a: callvirt instance void C2::set_P2(int32)
+            //   IL_001f: nop
+            //   IL_0020: stloc.0
+            //   IL_0021: ret
+            //
+            // We therefore visit the InitializedMember to get the implicit receiver for the contained initializer, and that implicit receiver will be cloned everywhere it encounters
+            // an IInstanceReferenceOperation with ReferenceKind InstanceReferenceKind.ImplicitReceiver
+
+            IOperation implicitReceiver = Visit(operation.InitializedMember);
+            HandleObjectOrCollectionInitializer(operation.Initializer, implicitReceiver);
+            return null;
+        }
+
+        private void HandleObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation initializer, IOperation implicitReceiver)
+        {
+            // PROTOTYPE(dataflow): Handle collection initializers
+            IOperation previousImplicitReceiver = _currentImplicitReceiver;
+            _currentImplicitReceiver = implicitReceiver;
+
+            foreach (IOperation innerInitializer in initializer.Initializers)
+            {
+                AddStatement(Visit(innerInitializer));
```

Comment:
>Visit(innerInitializer) [](start = 29, length = 23)

Rather than simply relying on default visit behavior, we probably should explicitly handle the initializers that we expect to run into, at least for success scenarios, and leave Visit for things we don't recognize.